nested query using virtual using_fts5 table

Hello

(full minimal example repo here : https://github.com/conradjones/sqlite_orm_fts_test/tree/main )


I'm having difficulty getting a query working on a normal table using a sub query (match) to select the IDs from the virtual fts5 table , to use to select the records in the primary normal table.

Its a contrived minimal example to demonstrate the point, 

Standard SQLite query would look like this
`SELECT * FROM normal_table where id in (SELECT normal_table_id FROM search_table WHERE text MATCH ?)`

Two tables / storage look like this 
```
struct NormalTable {
    int id;
    std::string text;
    int otherValue;
};

struct SearchTable {
    int normal_table_id;
    std::string text;
};

inline auto initStorage(const std::string &databaseFilePath) {
    return sqlite_orm::make_storage(
        databaseFilePath,
        sqlite_orm::make_table("normal_table",
                               sqlite_orm::make_column("id", &NormalTable::id,
                                                       sqlite_orm::primary_key().autoincrement()),
                               sqlite_orm::make_column("path", &NormalTable::text),
                               sqlite_orm::make_column("other_value", &NormalTable::otherValue)),

        sqlite_orm::make_virtual_table(
            "search_table",
            sqlite_orm::using_fts5(
                sqlite_orm::make_column("text", &SearchTable::text),
                sqlite_orm::make_column("normal_table_id", &SearchTable::normal_table_id)))
    );
}
````


any attempt to use them say like this 

```
auto rows = storage.iterate<NormalTable>(
        sqlite_orm::where(sqlite_orm::eq(&NormalTable::id,
                                         sqlite_orm::select(&SearchTable::normal_table_id,
                                                            sqlite_orm::where(
                                                                sqlite_orm::match<SearchTable>("Some Text"))))

        )


    );

 for (const auto &row: rows) {
    }
```

results in an error like 

```
error: no member named 'find_column_name' in 'sqlite_orm::internal::virtual_table_t<sqlite_orm::internal::using_fts5_t<db::SearchFile, sqlite_orm::internal::column_t<std::string db::SearchFile::*, sqlite_orm::internal::empty_setter>, sqlite_orm::internal::column_t<int db::SearchFile::*, sqlite_orm::internal::empty_setter>>>'
 12528 |             return pick_table<O>(dbObjects).find_column_name(field);
```

in sqlite_orm.h from the single include folder in dev branch 

around line 18544 

`isOpenedForever{connectionCtrl.open_forever},`

should be moved below 

` inMemory(filename.empty() || filename == ":memory:"),`

to prevent clang (and possibly other compilers) complaining about member initialization order.


> 
/Users/conrad/live_code/audio_app/third_party/sqlite_orm/include/sqlite_orm/sqlite_orm.h:18565:56: error: field 'isOpenedForever' will be initialized after field 'pragma' [-Werror,-Wreorder-ctor]
 18565 |                 on_open{std::move(onOpenSpec.onOpen)}, isOpenedForever{connectionCtrl.open_forever}, <

Thanks for this btw, 

The contrived example works fine with those changes, 

I ported the changes to my real project and it failed. I tracked it down to it using the tokenizer. 

I was able to repro in the contrived example by adding the tokenize option , I've updated the contrived example repo ( https://github.com/conradjones/sqlite_orm_fts_test/blob/main/main.cpp )

```
        sqlite_orm::make_virtual_table(
            "search_table",
            sqlite_orm::using_fts5(
                sqlite_orm::make_column("text", &SearchTable::text),
                sqlite_orm::make_column("normal_table_id", &SearchTable::normal_table_id)

                , sqlite_orm::tokenize("trigram")
            ))
    );
```

ends up with this error 

```
 In file included from /Users/conrad/CLionProjects/sqtest/main.cpp:2:
/Users/conrad/CLionProjects/sqtest/sqlite_orm.h:12653:47: error: no member named 'member_pointer' in 'sqlite_orm::internal::tokenize_t<const char *>'
 12653 |                     if (compare_fields(column.member_pointer, memberPointer) ||
       |                                        ~~~~~~ ^

```

--- a/.vscode/settings.json
+++ b/.vscode/settings.json
@@ -74,6 +74,18 @@
         "type_traits": "cpp",
         "utility": "cpp",
         "__config": "cpp",
-        "any": "cpp"
+        "any": "cpp",
+        "__threading_support": "cpp",
+        "complex": "cpp",
+        "condition_variable": "cpp",
+        "execution": "cpp",
+        "fstream": "cpp",
+        "future": "cpp",
+        "iostream": "cpp",
+        "numbers": "cpp",
+        "queue": "cpp",
+        "semaphore": "cpp",
+        "stack": "cpp",
+        "unordered_set": "cpp"
     }
 }

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -55,6 +55,12 @@ target_sources(sqlite_orm INTERFACE $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/incl
 
 target_include_directories(sqlite_orm INTERFACE $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>)
 
+# Enable reorder-ctor warning and make it an error
+target_compile_options(sqlite_orm INTERFACE
+    -Wreorder-ctor
+    -Werror=reorder-ctor
+)
+
 include(ucm)
 
 if (MSVC)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -55,10 +55,10 @@ target_sources(sqlite_orm INTERFACE $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/incl
 
 target_include_directories(sqlite_orm INTERFACE $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>)
 
-if (CMAKE_CXX_COMPILER_ID MATCHES "Clang|GNU")
+if(CMAKE_CXX_COMPILER_ID MATCHES "Clang|GNU")
     target_compile_options(sqlite_orm INTERFACE
-        -Wreorder-ctor
-        -Werror=reorder-ctor
+        -Wreorder
+        -Werror=reorder
     )
 endif()
 

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -55,11 +55,12 @@ target_sources(sqlite_orm INTERFACE $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/incl
 
 target_include_directories(sqlite_orm INTERFACE $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>)
 
-# Enable reorder-ctor warning and make it an error
-target_compile_options(sqlite_orm INTERFACE
-    -Wreorder-ctor
-    -Werror=reorder-ctor
-)
+if (CMAKE_CXX_COMPILER_ID MATCHES "Clang|GNU")
+    target_compile_options(sqlite_orm INTERFACE
+        -Wreorder-ctor
+        -Werror=reorder-ctor
+    )
+endif()
 
 include(ucm)
 

--- a/dev/schema/table.h
+++ b/dev/schema/table.h
@@ -401,12 +401,14 @@ namespace sqlite_orm {
                 using field_type = member_field_type_t<M>;
 
                 const std::string* res = nullptr;
-                iterate_tuple(this->columns, [&res, memberPointer](auto& column) {
-                    if (compare_fields(column.member_pointer, memberPointer) ||
-                        compare_fields(column.setter, memberPointer)) {
-                        res = &column.name;
-                    }
-                });
+                iterate_tuple(this->columns,
+                              col_index_sequence_with_field_type<columns_type, field_type>{},
+                              [&res, memberPointer](auto& column) {
+                                  if (compare_fields(column.member_pointer, memberPointer) ||
+                                      compare_fields(column.setter, memberPointer)) {
+                                      res = &column.name;
+                                  }
+                              });
                 return res;
             }
         };

--- a/dev/schema/table.h
+++ b/dev/schema/table.h
@@ -231,14 +231,16 @@ namespace sqlite_orm {
              *  @return column name or empty string if nothing found.
              */
             template<class M, satisfies<std::is_member_pointer, M> = true>
-            const std::string* find_column_name(M m) const {
-                const std::string* res = nullptr;
+            const std::string* find_column_name(M memberPointer) const {
                 using field_type = member_field_type_t<M>;
+
+                const std::string* res = nullptr;
                 iterate_tuple(this->elements,
                               col_index_sequence_with_field_type<elements_type, field_type>{},
-                              [&res, m](auto& c) {
-                                  if (compare_fields(c.member_pointer, m) || compare_fields(c.setter, m)) {
-                                      res = &c.name;
+                              [&res, memberPointer](auto& column) {
+                                  if (compare_fields(column.member_pointer, memberPointer) ||
+                                      compare_fields(column.setter, memberPointer)) {
+                                      res = &column.name;
                                   }
                               });
                 return res;
@@ -343,6 +345,11 @@ namespace sqlite_orm {
             void for_each_column(L&& lambda) const {
                 this->module_details.for_each_column(lambda);
             }
+
+            template<class MP, satisfies<std::is_member_pointer, MP> = true>
+            const std::string* find_column_name(MP memberPointer) const {
+                return this->module_details.find_column_name(memberPointer);
+            }
         };
 
         template<class T>
@@ -387,6 +394,20 @@ namespace sqlite_orm {
             void for_each_column(L&& lambda) const {
                 using col_index_sequence = filter_tuple_sequence_t<columns_type, is_column>;
                 iterate_tuple(this->columns, col_index_sequence{}, lambda);
+            }
+
+            template<class M, satisfies<std::is_member_pointer, M> = true>
+            const std::string* find_column_name(M memberPointer) const {
+                using field_type = member_field_type_t<M>;
+
+                const std::string* res = nullptr;
+                iterate_tuple(this->columns, [&res, memberPointer](auto& column) {
+                    if (compare_fields(column.member_pointer, memberPointer) ||
+                        compare_fields(column.setter, memberPointer)) {
+                        res = &column.name;
+                    }
+                });
+                return res;
             }
         };
 #endif

--- a/dev/storage_base.h
+++ b/dev/storage_base.h
@@ -667,10 +667,9 @@ namespace sqlite_orm {
                          connection_control connectionCtrl,
                          on_open_spec onOpenSpec,
                          int foreignKeysCount) :
-                on_open{std::move(onOpenSpec.onOpen)}, isOpenedForever{connectionCtrl.open_forever},
-                pragma(std::bind(&storage_base::get_connection, this)),
+                on_open{std::move(onOpenSpec.onOpen)}, pragma(std::bind(&storage_base::get_connection, this)),
                 limit(std::bind(&storage_base::get_connection, this)),
-                inMemory(filename.empty() || filename == ":memory:"),
+                inMemory(filename.empty() || filename == ":memory:"), isOpenedForever{connectionCtrl.open_forever},
                 connection(std::make_unique<connection_holder>(
                     std::move(filename),
                     std::bind(&storage_base::on_open_internal, this, std::placeholders::_1))),

--- a/include/sqlite_orm/sqlite_orm.h
+++ b/include/sqlite_orm/sqlite_orm.h
@@ -12649,12 +12649,14 @@ namespace sqlite_orm {
                 using field_type = member_field_type_t<M>;
 
                 const std::string* res = nullptr;
-                iterate_tuple(this->columns, [&res, memberPointer](auto& column) {
-                    if (compare_fields(column.member_pointer, memberPointer) ||
-                        compare_fields(column.setter, memberPointer)) {
-                        res = &column.name;
-                    }
-                });
+                iterate_tuple(this->columns,
+                              col_index_sequence_with_field_type<columns_type, field_type>{},
+                              [&res, memberPointer](auto& column) {
+                                  if (compare_fields(column.member_pointer, memberPointer) ||
+                                      compare_fields(column.setter, memberPointer)) {
+                                      res = &column.name;
+                                  }
+                              });
                 return res;
             }
         };
@@ -18562,10 +18564,9 @@ namespace sqlite_orm {
                          connection_control connectionCtrl,
                          on_open_spec onOpenSpec,
                          int foreignKeysCount) :
-                on_open{std::move(onOpenSpec.onOpen)}, isOpenedForever{connectionCtrl.open_forever},
-                pragma(std::bind(&storage_base::get_connection, this)),
+                on_open{std::move(onOpenSpec.onOpen)}, pragma(std::bind(&storage_base::get_connection, this)),
                 limit(std::bind(&storage_base::get_connection, this)),
-                inMemory(filename.empty() || filename == ":memory:"),
+                inMemory(filename.empty() || filename == ":memory:"), isOpenedForever{connectionCtrl.open_forever},
                 connection(std::make_unique<connection_holder>(
                     std::move(filename),
                     std::bind(&storage_base::on_open_internal, this, std::placeholders::_1))),

--- a/include/sqlite_orm/sqlite_orm.h
+++ b/include/sqlite_orm/sqlite_orm.h
@@ -12479,14 +12479,16 @@ namespace sqlite_orm {
              *  @return column name or empty string if nothing found.
              */
             template<class M, satisfies<std::is_member_pointer, M> = true>
-            const std::string* find_column_name(M m) const {
-                const std::string* res = nullptr;
+            const std::string* find_column_name(M memberPointer) const {
                 using field_type = member_field_type_t<M>;
+
+                const std::string* res = nullptr;
                 iterate_tuple(this->elements,
                               col_index_sequence_with_field_type<elements_type, field_type>{},
-                              [&res, m](auto& c) {
-                                  if (compare_fields(c.member_pointer, m) || compare_fields(c.setter, m)) {
-                                      res = &c.name;
+                              [&res, memberPointer](auto& column) {
+                                  if (compare_fields(column.member_pointer, memberPointer) ||
+                                      compare_fields(column.setter, memberPointer)) {
+                                      res = &column.name;
                                   }
                               });
                 return res;
@@ -12591,6 +12593,11 @@ namespace sqlite_orm {
             void for_each_column(L&& lambda) const {
                 this->module_details.for_each_column(lambda);
             }
+
+            template<class MP, satisfies<std::is_member_pointer, MP> = true>
+            const std::string* find_column_name(MP memberPointer) const {
+                return this->module_details.find_column_name(memberPointer);
+            }
         };
 
         template<class T>
@@ -12635,6 +12642,20 @@ namespace sqlite_orm {
             void for_each_column(L&& lambda) const {
                 using col_index_sequence = filter_tuple_sequence_t<columns_type, is_column>;
                 iterate_tuple(this->columns, col_index_sequence{}, lambda);
+            }
+
+            template<class M, satisfies<std::is_member_pointer, M> = true>
+            const std::string* find_column_name(M memberPointer) const {
+                using field_type = member_field_type_t<M>;
+
+                const std::string* res = nullptr;
+                iterate_tuple(this->columns, [&res, memberPointer](auto& column) {
+                    if (compare_fields(column.member_pointer, memberPointer) ||
+                        compare_fields(column.setter, memberPointer)) {
+                        res = &column.name;
+                    }
+                });
+                return res;
             }
         };
 #endif

--- a/tests/schema/virtual_table.cpp
+++ b/tests/schema/virtual_table.cpp
@@ -23,7 +23,7 @@ TEST_CASE("virtual table") {
     auto virtualTable =
         make_virtual_table("posts", using_fts5(make_column("title", &Post::title), make_column("body", &Post::body)));
     {
-        auto compareColumnName = [](const std::string* foundValue, std::string expectedValue) {
+        const auto compareColumnName = [](const std::string* foundValue, std::string expectedValue) {
             if (!foundValue) {
                 return false;
             }

--- a/tests/schema/virtual_table.cpp
+++ b/tests/schema/virtual_table.cpp
@@ -20,11 +20,22 @@ TEST_CASE("virtual table") {
 #endif
     };
 
+    auto virtualTable =
+        make_virtual_table("posts", using_fts5(make_column("title", &Post::title), make_column("body", &Post::body)));
+    {
+        auto compareColumnName = [](const std::string* foundValue, std::string expectedValue) {
+            if (!foundValue) {
+                return false;
+            }
+            return *foundValue == expectedValue;
+        };
+        REQUIRE(compareColumnName(virtualTable.find_column_name(&Post::title), std::string("title")));
+        REQUIRE(compareColumnName(virtualTable.find_column_name(&Post::body), std::string("body")));
+    }
+
     /// CREATE VIRTUAL TABLE posts
     /// USING FTS5(title, body);
-    auto storage = make_storage(
-        "",
-        make_virtual_table("posts", using_fts5(make_column("title", &Post::title), make_column("body", &Post::body))));
+    auto storage = make_storage("", std::move(virtualTable));
 
     storage.sync_schema();
     storage.sync_schema_simulate();
@@ -79,5 +90,35 @@ TEST_CASE("virtual table") {
         storage.select(columns(highlight<Post>(0, "<b>", "</b>"), highlight<Post>(1, "<b>", "</b>")),
                        where(match<Post>("SQLite")),
                        order_by(rank()));
+}
+
+TEST_CASE("issue1410") {
+    struct NormalTable {
+        int id;
+        std::string text;
+        int otherValue;
+    };
+
+    struct SearchTable {
+        int normal_table_id;
+        std::string text;
+    };
+
+    auto storage =
+        make_storage("",
+                     make_table("normal_table",
+                                make_column("id", &NormalTable::id, primary_key().autoincrement()),
+                                make_column("path", &NormalTable::text),
+                                make_column("other_value", &NormalTable::otherValue)),
+
+                     make_virtual_table("search_table",
+                                        using_fts5(make_column("text", &SearchTable::text),
+                                                   make_column("normal_table_id", &SearchTable::normal_table_id))));
+    storage.sync_schema();
+    auto rows = storage.iterate<NormalTable>(
+        where(eq(&NormalTable::id, select(&SearchTable::normal_table_id, where(match<SearchTable>("Some Text"))))));
+
+    for (const auto& row: rows) {
+    }
 }
 #endif

--- a/tests/schema/virtual_table.cpp
+++ b/tests/schema/virtual_table.cpp
@@ -113,12 +113,13 @@ TEST_CASE("issue1410") {
 
                      make_virtual_table("search_table",
                                         using_fts5(make_column("text", &SearchTable::text),
-                                                   make_column("normal_table_id", &SearchTable::normal_table_id))));
+                                                   make_column("normal_table_id", &SearchTable::normal_table_id),
+                                                   tokenize("trigram"))));
     storage.sync_schema();
     auto rows = storage.iterate<NormalTable>(
         where(eq(&NormalTable::id, select(&SearchTable::normal_table_id, where(match<SearchTable>("Some Text"))))));
 
     for (const auto& row: rows) {
-    }
+    }  //  has to be compiled
 }
 #endif

